write-fonts/src/graph.rs

```diff
@@ -408,157 +409,208 @@ impl Graph {
         self.distance_invalid = false;
     }
 
-    /// Returns `true` if there were any 32bit subgraphs
-    fn assign_32bit_spaces(&mut self) -> bool {
+    /// isolate and assign spaces to subgraphs reachable via long offsets.
+    ///
+    /// This finds all subgraphs that are reachable via long offsets, and
+    /// isolates them (ensuring they are *only* reachable via long offsets),
+    /// assigning each unique space an identifier.
+    ///
+    /// Each space may have multiple roots; this works by finding the connected
+    /// components from each root (counting only nodes reachable via long offsets).
+    fn assign_spaces_hb(&mut self) -> bool {
         self.update_parents();
-        // find all the nodes that have incoming 32-bit edges
-        let mut roots = HashSet::new();
-        for (id, node) in &self.nodes {
-            if node
-                .parents
-                .iter()
-                .any(|(_, len)| *len == OffsetLen::Offset32)
-            {
-                roots.insert(*id);
-            }
-        }
+        let (visited, mut roots) = self.find_space_roots_hb();
 
         if roots.is_empty() {
             return false;
         }
 
-        // assign all nodes reachable from 16/24 bit edges to space 0.
-        self.assign_space_0();
+        log::debug!("found {} space roots to isolate", roots.len());
 
-        while !roots.is_empty() {
-            let root = *roots.iter().next().unwrap();
-            self.isolate_and_assign_space(root);
-            roots.remove(&root);
+        // we want to *invert* the visited set, but we don't have a fancy hb_set_t
```

Comment:
should we file an issue to write an hb_set_t? - I believe some of hb's structures are very carefully tuned to work well for text.

Reply:
We've got one (#192)